### PR TITLE
Buffer metrics before writing to GCS/Tensorboard

### PR DESCRIPTION
**Pull request by SurbhiJainUSC**
# Description

This PR adjusts how we calculate step time, reverting to the method used before https://github.com/AI-Hypercomputer/maxtext/pull/1815. Our previous attempt to precisely measure step time involved adding `jax.block_until_ready()`, which, while accurate for individual steps, introduced an undesirable delay between `train_steps` during profiling. This delay stemmed from JAX being prevented from queuing up the next training step. To improve training performance, we're now prioritizing the ability for JAX to queue steps efficiently, even if it means a slight compromise in the precision of individual step time reports.

# Tests

```
completed step: 0, seconds: 18.665, TFLOP/s/device: 2.263, Tokens/s/device: 54.862, total_weights: 3573, loss: 0.993
completed step: 1, seconds: 9.720, TFLOP/s/device: 4.346, Tokens/s/device: 105.355, total_weights: 3253, loss: 0.993
completed step: 2, seconds: 0.164, TFLOP/s/device: 256.874, Tokens/s/device: 6226.779, total_weights: 1692, loss: 0.912
completed step: 3, seconds: 0.834, TFLOP/s/device: 50.645, Tokens/s/device: 1227.665, total_weights: 2944, loss: 1.105
completed step: 4, seconds: 0.957, TFLOP/s/device: 44.151, Tokens/s/device: 1070.256, total_weights: 3466, loss: 0.948
completed step: 5, seconds: 0.961, TFLOP/s/device: 43.949, Tokens/s/device: 1065.339, total_weights: 2420, loss: 0.946
completed step: 6, seconds: 16.063, TFLOP/s/device: 2.630, Tokens/s/device: 63.748, total_weights: 1655, loss: 1.058
completed step: 7, seconds: 0.013, TFLOP/s/device: 3333.582, Tokens/s/device: 80808.081, total_weights: 3335, loss: 1.000
completed step: 8, seconds: 0.964, TFLOP/s/device: 43.820, Tokens/s/device: 1062.227, total_weights: 1676, loss: 0.850
completed step: 9, seconds: 0.962, TFLOP/s/device: 43.895, Tokens/s/device: 1064.035, total_weights: 1676, loss: 0.965
```

Gap between two `jit_train_steps` reduced from 30ms 
image

to 13.8us 
image

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/metric_logger.py [hunk 1]
```diff
@@ -153,8 +154,17 @@ def get_performance_metric_queue(self, config):
         gcp_workload_monitor.start_performance_reporting_thread(performance_metric_queue)
     return performance_metric_queue
 
+  def buffer_and_write_train_metrics(self, metrics, step, step_time_delta):
```
Comment (gobbleturk) [inline MaxText/metric_logger.py:157]: Can you add this buffer approach allows train steps to be executed back to back, in particular allows overlapping the data loading of step n during the execution of step n-1

Comment (SurbhiJainUSC) [reply MaxText/metric_logger.py:157]: Updated.

File: MaxText/elastic_train.py [hunk 2]
```diff
@@ -225,17 +226,17 @@ def train_loop(config, elastic_manager, recorder, state=None):
...
+        last_step_completion = datetime.datetime.now()
+
         checkpointing.maybe_save_checkpoint(checkpoint_manager, state, config, data_iterator, step)
 
         prof.maybe_deactivate_profiler(step, state)
 
       if step == start_step:
         max_utils.print_mem_stats("After params initialized")
 
-      jax.block_until_ready(state)  # ensure training step is completed
```
Comment (shauryagup) [inline MaxText/elastic_train.py:235]: After removing this, wouldn't the step time be simply the time to dispatch the step and not its actual execution time? This behavior could also result in different step times between pathways and mcjax at least on the JAX layer since both of them use different dispatch strategies. 

Do we know what was python doing during the 30 ms reported in the first profile? Was it only spent on dispatching the operation? Or does it also include data loading etc.?

File: MaxText/metric_logger.py [hunk 3]
```diff
@@ -153,8 +154,22 @@ def get_performance_metric_queue(self, config):
...
```
Comment (lukebaumann) [inline MaxText/metric_logger.py:157]: Can you optionally block on the metrics so that the user can choose to measure step time (dispatch N to finish N) and have a simpler flow (avoiding buffering of metrics and queuing of steps) vs measuring step time (dispatch N to dispatch N+1) and added complexity?

File: MaxText/elastic_train.py (hunk 2 above)
Comment (SurbhiJainUSC) [reply MaxText/elastic_train.py:235]: https://github.com/AI-Hypercomputer/maxtext/pull/1815 was actually fixing the issue with reported step time. Before #1815, we were taking into account the time it takes for step n-1 to complete when reporting step time for step n. With #1815, I tried fixing this issue by ensuring that step time for step n only includes the actual time it takes to complete step n and not step n-1. This was achieved by adding `jax.block_until_ready()`. However, this blocks JAX from queuing any operation until the current training step is completed.

In the first profile, 30ms include the time for data loading for step n before starting training step.

There is a tradeoff between accuracy (reporting the correct step time) vs performance here.

File: MaxText/metric_logger.py [hunk 4]
```diff
@@ -153,8 +154,22 @@ def get_performance_metric_queue(self, config):
...
+    """
+    Buffers metrics for the current training step and simultaneously writes the training metrics
+    for the previous step to GCS and/or TensorBoard. This buffering strategy allows for back-to-back
+    execution of training steps, by overlapping data loading for step n with the execution of step n−1.
+    This significantly boosts training efficiency.
```
Comment (lukebaumann) [inline MaxText/metric_logger.py:162]: IIUC in the example step times were about 0.9 seconds and this saves (32 ms - 13 us) ~ 0.030 seconds or 3.3% difference.

Should be changed to: "This boosts training efficiency by a few percent" since "significant" is subjective.

File: MaxText/elastic_train.py (hunk 2 above)
Comment (lukebaumann) [reply MaxText/elastic_train.py:235]: > Do we know what was python doing during the 30 ms reported in the first profile? Was it only spent on dispatching the operation? Or does it also include data loading etc.?

As an alternative to this PR's approach, if the loop is reordered so that data loading for the next step happens at the end of the previous iteration and the block on metrics happens after the data loading is dispatched, you may achieve similar performance without needed to buffer the metrics.

File: MaxText/metric_logger.py [hunk 5]
```diff
@@ -57,6 +57,7 @@ def __init__(self, config, learning_rate_schedule):
     self.performance_metric_queue = self.get_performance_metric_queue(config)
     self.learning_rate_schedule = learning_rate_schedule
     self.cumulative_eval_metrics = {"scalar": defaultdict(float)}
+    self.buffered_train_metrics = None
```
Comment (lukebaumann) [inline MaxText/metric_logger.py:60]: AFAICT the buffered metric from before a failure may be incomplete but because `elastic_handler` creates a brand new `MetricLogger` there is no change needed.

**Review by lukebaumann — COMMENTED** [review]
I will run some elastic training runs from this branch and let you know if they pass.

File: MaxText/metric_logger.py (hunk 3 above)
Comment (SurbhiJainUSC) [reply MaxText/metric_logger.py:157]: IMO, configuring this will add too much complexity in the code on top of another config variable (which would be a different pain to handle).

File: MaxText/elastic_train.py (hunk 2 above)
Comment (lukebaumann) [reply MaxText/elastic_train.py:235]: We can discuss the alternative as a potential future improvement if you are interested.

> Do we know what was python doing during the 30 ms reported in the first profile? Was it only spent on dispatching the operation? Or does it also include data loading etc.?
I think this is unanswered still so I will let @shauryagup resolve the conversation when he is satisfied.

File: MaxText/metric_logger.py (hunk 5 above)
Comment (lukebaumann) [reply MaxText/metric_logger.py:60]: Confirmed

File: MaxText/metric_logger.py (hunk 3 above)
Comment (lukebaumann) [reply MaxText/metric_logger.py:157]: I think there is a simple implementation of my idea but that discussion won't block this PR.

File: MaxText/metric_logger.py (hunk 4 above)
Comment (lukebaumann) [reply MaxText/metric_logger.py:162]: Depending on the configuration, the savings will change but I would be curious if it is ever more than a "few percent".

This suggestion also shouldn't block this PR so I am resolving.

**Comment by lukebaumann** [comment]
> I will run some elastic training runs from this branch and let you know if they pass.

I ran elastic training from this branch on 2 slices and kill each slice at different times for a total of three disruptions. The training reacted exactly as expected.

Thanks!

**Review by lukebaumann — APPROVED** [review]

**Review by khatwanimohit — APPROVED** [review]
